### PR TITLE
Implementing `/lts/state/account-deposit-behaviour` endpoint

### PR DESCRIPTION
**Pull request by jakrawcz-rdx**
Addresses https://whimsical.com/babylon-work-tracker-Am2s7hSa7xHd5yCxgub67a@2bsEvpTYSt1Hj3SDQYhbvr4oNy2rQseEsc6

More detailed spec agreed on Slack:
> New “lts/state/account-deposit-behaviour” takes account address, list of resource addresses (0 ≤ x ≤ 20), and optional badge for the deposit.
> Returns:
> - General account resource preference
> - Whether the badge is approved for deposits
> - For each requested resource address:
>    - What its overriden behaviour is (if at all)
>    - Whether the vault exists already for this resource
>    - [Whether the resource is XRD]
>    - The resolved ability for whether a deposit of that resource to that account with that badge can succeed.
> - ~~Whether all requested deposits resolve successfully - ie whether the deposit instruction will succeed (or abort/refund which it does atomically)~~

> Background reading: https://docs-babylon.radixdlt.com/main/scrypto/native-blueprints/accounts.html#_configuring_account_deposit_modes_and_resource_preference_map (edited)

### CODE REVIEW

File: core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        badge_of_authorized_depositor,
+        resource_specific_behaviours,
+    })
+    .map(Json)
+}
+
+/// Resolves whether the deposit is allowed, based on raw inputs.
+/// See https://docs-babylon.radixdlt.com/main/scrypto/native-blueprints/accounts.html for a
+/// flow chart.
+fn allows_deposit(
```
Comment (jakrawcz-rdx) [inline core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs:221]: _(note from author)_

## Dear reviewers,

Please pay special attention to this method.

**Review by LukasGasior1 — APPROVED** [review]
Looks good as far as I can tell, but maybe someone from the builders team could double check the `allows_deposit` logic (@talekhinezh , @iamyulong ?)

File: core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs (hunk 1 above)
Comment (iamyulong) [reply core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs:221]: Maybe rename to `allows_try_deposit` to be consistent with the blueprint?

File: core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        Some(AccountResourcePreference::Disallowed) => badge_of_authorized_depositor == &Some(true),
+        // No explicit resource preference - try the same logic with the default deposit rule:
+        None => {
+            match default_deposit_rule {
+                // The default rule allows it:
+                DefaultDepositRule::Accept => true,
+                // The default rule disallows it, but a present authorized badge can override this:
+                DefaultDepositRule::Reject => badge_of_authorized_depositor == &Some(true),
+                // The extra case of the default rule, allowing existing vaults *or* XRD:
+                DefaultDepositRule::AllowExisting => vault_exists || is_xrd,
```
Comment (iamyulong) [inline core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs:241]: I think this is incorrect. 

Should be `vault_exists || is_xrd || badge_of_authorized_depositor == &Some(true)`?

Maybe @0xOmarA can confirm :-D

Comment (jakrawcz-rdx) [reply core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs]: OMG you are right, and even the flow chart in the official docs agrees 😱 

Basically it looks like presenting a badge of authorized depositor overrides everything, so I now just moved this condition first, PTAL.

File: core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs (hunk 1 above)
Comment (iamyulong) [reply core-rust/core-api-server/src/core_api/handlers/lts/state_account_deposit_behaviour.rs:221]: It's not a big deal. Pick whichever makes more sense within your context. :)

**Comment by iamyulong** [comment]
The updated `allows_deposit` logic looks good to me.

**Review by iamyulong — APPROVED** [review]
I only reviewed the `allows_deposit` logic. :-D

File: core/src/test/java/com/radixdlt/api/core/LtsAccountDepositBehaviourTest.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      assertThat(differentResult.getResourceSpecificBehaviours())
+          .isEqualTo(
+              Map.of(
+                  addressing.encode(ScryptoConstants.XRD_RESOURCE_ADDRESS),
+                  new ResourceSpecificDepositBehaviour()
+                      .resourcePreference(null)
+                      .vaultExists(true)
+                      .isXrd(true)
+                      .depositAllowed(true)));
+    }
```
Comment (dhedey) [inline core/src/test/java/com/radixdlt/api/core/LtsAccountDepositBehaviourTest.java:152]: I'm slightly worried we haven't tested that they keys and payloads are serialized correctly from the stores.

I suggest the following two test cases:
* Changes default behaviour to DenyAll & a test which checks if we can submit XRD - this is most important test for the exchange use case.
* Sets default behaviour to Existing, set an authorized depositor badge and a resource preference on the account, and check that the response is correct 

We'll need to add some manifests to `Manifest.java` for these.

And then can submit those manifests with eg:
```java
      var accountKeyPair = ECKeyPair.generateNew();
      var accountAddress = Address.virtualAccountAddress(accountKeyPair.getPublicKey());
      submitAndWaitForSuccess(test, Manifest.zYZ(accountAddress), List.of(accountKeyPair));
```

**Review by dhedey — APPROVED** [review]
Really nice. A few minor renames / additional tests suggestions. Please fix, add tests and merge into develop when they pass.

File: core-rust/core-api-server/core-api-schema.yaml [hunk 4]
```diff
@@ -5544,6 +5603,82 @@ components:
...
+      properties:
+        state_version:
+          $ref: "#/components/schemas/StateVersion"
+          description: The state version at which the query was performed.
+        ledger_header_summary:
+          $ref: "#/components/schemas/LedgerHeaderSummary"
+          description: The excerpt from the ledger header committed at the `state_version`.
+        default_deposit_rule:
+          $ref: '#/components/schemas/DefaultDepositRule'
+        badge_of_authorized_depositor:
```
[the remark this answers is not in the document]
Comment (dhedey) [reply core-rust/core-api-server/core-api-schema.yaml:5644]: (Just a general principle that I prefer booleans to have verbs)

File: core/src/test/java/com/radixdlt/api/core/LtsAccountDepositBehaviourTest.java (hunk 3 above)
Comment (jakrawcz-rdx) [reply core/src/test/java/com/radixdlt/api/core/LtsAccountDepositBehaviourTest.java:152]: Added 👍 
(almost exactly the 2 scenarios you describe; anyway I am sure that they verify that "the right resource preference is read from the KV" and "the right depositor badge is read from the KV")

**Comment by jakrawcz-rdx** [comment]
I only see 1 failure (the known `MultiNodeRebootTest restart_all_nodes_intermittently[2]` instability), hence I consider the CI 🟢 ; merging!

**Pull request merged**